keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2144 @@
+controller. For a simple drain, cluster admin can simply create the NodeMaintenance with
+`stage: Drain` directly.
+
+### DaemonSet Controller
+
+The DaemonSet workloads should be tied to the node lifecycle because they typically run critical
+workloads where availability is paramount. Therefore, the daemonset controller should respond to
+the Evacuation only if there is a NodeMaintenance happening on that node and the DaemonSet is in
+the `drainTargets`. For example, if we observe the following NodeMaintenance:
+
+```yaml
+apiVersion: v1alpha1
+kind: NodeMaintenance
+...
+status:
+  nodeStatuses:
+    - nodeRef:
+        name: six
+      drainTargets:
+        - podPriority: 2147483647
+          podType: Default
+        - podPriority: 5000
+          podType: DaemonSet
+  ...
+```
+
+To register the daemonset controller as a controller evacuator, it should set the following
+annotation on its own pods.
+
+```yaml
+evacuation.coordination.k8s.io/daemonset.apps.k8s.io: "10000/controller"
+```
+
+The controller should respond to the Evacuation object when it observes its own class
+(`daemonset.apps.k8s.io`) in `.status.activeEvacuatorClass`.
+
+For the above node maintenance, the controller should not react to Evacuations of DaemonSet pods
+with a priority greater than 5000. This state should not normally occur, as Evacuation requests
+should be coordinated with NodeMaintenance. If it does occur, we should not encourage this flow by
+updating the `.status.ActiveEvacuatorCompleted` field, although it is required to update this field
+for normal workloads.
+
+If the DaemonSet pods have a priority equal to or less than 5000, the Evacuation status should be
+updated appropriately as follows, and the targeted pod should be deleted by the daemonset
+controller:
+
+```yaml
+apiVersion: v1alpha1
+kind: Evacuation
+metadata:
+  finalizers:
+    evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io
+  labels:
+    app: critical-ds
+  name: ae9b4bc6-e4ca-4f8e-962b-2d4459b1f684-critical-ds-5nxjs
+  namespace: critical-workloads
+spec:
+  podRef:
+    name: critical-ds-5nxjs
+    uid:  ae9b4bc6-e4ca-4f8e-962b-2d4459b1f684
+  progressDeadlineSeconds: 1800
+  evacuators:
+    - evacuatorClass: daemonset.apps.k8s.io
+      priority: 10000
+      role: controller
+status:
+  activeEvacuatorClass: daemonset.apps.k8s.io
+  activeEvacuatorCompleted: false
+  evacuationProgressTimestamp: "2024-04-22T21:40:32Z"
+  expectedEvacuationFinishTime: "2024-04-22T21:41:32Z" # now + terminationGracePeriodSeconds:
+  failedEvictionCounter: 0
+  message: "critical-ds is terminating the pod due to node maintenance (OS upgrade)."
+  conditions: []
+```
+
+Once the pod is terminated and removed from the node, it should not be re-scheduled on the node by
+the deamonset controller until the node maintenance is complete.
+
+### Kubelet: Graceful Node Shutdown
+
+The current Graceful Node Shutdown feature has a couple of drawbacks when compared to
+NodeMaintenance:
```

Comment:
graceful/non-graceful shutdown and NodeMaintenance procedures on a node could interfere.
As suggested [previously](https://github.com/kubernetes/enhancements/pull/4213#pullrequestreview-1971258767), NMO has achieved mutual exclusion from NHC (and node remediators) by using node leases that could hint who is doing remediation/maintenance so they both won't happen simultaneously and be agnostic.

These two procedures are important for a faulty node, thus it should be considered how they would work when a user wants them together:

- Blocking one of the procedures while one is running (/Active)
- Block and wait until one of the procedures is finished (so no interference)
- Modify the procedures so they can work together (sequentially) without interference

Reply:
@razo7 When I look at the NMO and NHC, the important part they solve is the node lifecycle/remediation (among other things).

I am not sure if NodeMaintenance's responsibility should be the node lifecycle. To check if the node is healthy or down. If the upgrade is in progress or not. If the shutdown is in progress or has finished. Etc.

Also, automatically triggering the shutdown after the node has drained, seems like a tricky topic.

I think it would be better to leave this responsibility to other actors (e.g. node lifecycle components) that can use the NodeMaintenance and then react to its status accordingly. One of these components is also kubelet, as described in the KEP. 

Initially, NodeMaintenance should focus on properly solving the node drain. To tell the applications how they should leave the node. In what order and safely.

We can discuss supporting additional features by NodeMaintenance later, after some adoption, if needed.

Reply:
Regarding the race for the node drain. NodeMaintenance should potentially solve all the important use cases.
And having multiple NodeMaintenance objects from different actors is already supported in the KEP.

Also, since this will be the most graceful solution for the node drain, we should not clash with other more aggressive drainers. (*although NodeMaintenance can be more aggressive for workloads guarded by PDBs).

Today's different types of evictions, deletions and drainers also do not synchronize who is removing what workload and when.

Reply:
For me, NodeMaintenance would provide three things:
- a coordination point for clients of the Kubernetes API to understand more about why a node might be tainted, out of service, etc
- a building block for additional, out-of-project controllers and operators
  - for example, make an owned NodeMaintenance that belongs to your ClusterKernelPatchRequest custom resource.
  - for example, automatically create a NodeMaintenance in response to an external API marking the **antarctica-d02** zone as down, matching all affected nodes by label. Remove it once you've sailed over and checked the cluster.
- server-side tracking and control for drains, `kubectl drain` style


We shouldn't try to provide too much of the specifics, so long as the basic thing doesn't make those use cases particularly hard to implement one day in the future.

Reply:
I can imagine a NodeMaintenance initiated as a result of a UPS reporting it has at most 3s of power left (because the UPS is overheating). I can also imagine scheduling node maintenance months in advance. There are more user stories than we have room to track.

Reply:
I have added these under the non goals: https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#non-goals to at least track this somewhere.

Also see #8 in [Motivation](https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#motivation) as @kwilczynski noted in https://github.com/kubernetes/enhancements/pull/4213#discussion_r1598305160. There are problems today, with the pod termination synchronization.